### PR TITLE
code review branch

### PR DESCRIPTION
**Pull request by hDoubleE**
Started from clean slate in master, should work better.

### CODE REVIEW

File: Palindrome.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+                    end--;
+                }
+                else
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+        public char[] PalindromeHelper(string s)
```
Comment (michsco) [inline Palindrome.cs]: A couple of items to consider about this method (some apply to interviewing and others just to regular coding):

1. Why is it public? Do you want your users to call it?
2. The method name should do a better job describing what you are doing here. Perhaps something like "ConvertToCharArray" or "ConvertToCleanCharArray".

Comment (michsco) [inline Palindrome.cs]: A couple of items to consider about this method (some apply to interviewing and others just to regular coding):

1. Why is it public? Do you want your users to call it?
2. The method name should do a better job describing what you are doing here. Perhaps something like "ConvertToCharArray" or "ConvertToCleanCharArray".

File: Palindrome.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        {
+            // Tests failed on Names and Sentences due to white space.
+            // Had to use String.Replace and ToLower methods in helper
+            // to fix input...is this right?
+            // Is pre-gaming the input cheating...or does the method need work?
+            string fixString = s.ToLower().Replace(" ", String.Empty);
```
Comment (michsco) [inline Palindrome.cs]: I really don't understand the purpose of this code. Is the idea that you want to restrict input from not allowing spaces? If that's the issue, I would recommend that you check for a space character and just throw an `ArgumentException` in that case.

File: Palindrome.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+namespace PalindromeArray
+{
+    public class Palindrome
+    {
+        public bool IsArrayPalindrome(char[] arr)
+        {
+            if (arr.Length == 0)
+            {
+                return false;
```
Comment (michsco) [inline Palindrome.cs]: You may want to add a comment here indicating that you are defining an empty string as not a palindrome.

File: Palindrome.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
+﻿using System;
+
+namespace PalindromeArray
+{
+    public class Palindrome
```
Comment (michsco) [inline Palindrome.cs]: Seal your class.

File: Palindrome.cs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        public bool IsArrayPalindrome(char[] arr)
```
Comment (michsco) [inline Palindrome.cs]: For actual software engineering projects, I tend to add XML comments (type /// right before the method and VS will help you) to help describe the behavior.

File: Palindrome.cs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        {
+            if (arr.Length == 0)
```
Comment (michsco) [inline Palindrome.cs]: You forgot to check for the `arr` being null. In that case, you should throw `ArgumentNullException(nameof(arr))`.

File: Palindrome.cs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+            {
+                return false;
+            }
+
+            int start = 0;
+            int end = arr.Length - 1;
+
+            while (start <= end)
```
Comment (michsco) [inline Palindrome.cs]: You don't need the `=` here as the code will behave normally.

File: Palindrome.UnitTests.cs [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+// Added a test for Empty Input.
+// Created branch 'review' for PR sharing.
+
+
+namespace PalindromeArray.UnitTests
+{
+    [TestClass]
+    public class PalindromeUnitTests
```
Comment (michsco) [inline Palindrome.UnitTests.cs]: Seal

File: Palindrome.UnitTests.cs [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+            bool boolResultOfPalindromeCheck = check.IsArrayPalindrome(arrToCheck);
+
+            Assert.AreEqual(false, boolResultOfPalindromeCheck);
+        }
+        [TestMethod]
+        public void Check_Sentence_Palindrome()
+        {
+            Palindrome check = new Palindrome();
+            char[] arrToCheck = check.PalindromeHelper("Murder for a jar of red rum");
```
Comment (michsco) [inline Palindrome.UnitTests.cs]: If you want to support this scenario, I would update the `IsArrayPalindrome` method to accept a `string` with code like this:

```csharp
public bool IsArrayPalindrome(string input)
{
    if (string.isNullOrWhitespace(input))
    {
        throw new ArgumentNullException(nameof(input));
    }
    
    char[] values = input.Trim().Replace(" ", string.Empty).ToCharArray();
    ...
}
```
You don't really need your helper method.

File: Palindrome.UnitTests.cs (hunk 8 above)
Comment (michsco) [inline Palindrome.UnitTests.cs]: You are missing a test case for null.

File: Palindrome.UnitTests.cs [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+{
+    [TestClass]
+    public class PalindromeUnitTests
+    {
+        [TestMethod]
+        public void Check_Empty_Input()
+        {
+            // Arrange
+            Palindrome check = new Palindrome();
+            char[] arrToCheck = check.PalindromeHelper("");
```
Comment (michsco) [inline Palindrome.UnitTests.cs]: This is a general comment about the helper. It seems like the `PalindromeHelper` method is only being used from tests, not from production code, so you should move it into the test class itself.

File: Palindrome.UnitTests.cs [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+            // Act
+            bool boolResultOfPalindromeCheck = check.IsArrayPalindrome(arrToCheck);
+            // Assert
+            Assert.AreEqual(false, boolResultOfPalindromeCheck);
+        }
+        [TestMethod]
+        public void Check_Longest_Palindrome_In_Dictionary()
+        {
+            Palindrome check = new Palindrome();
```
Comment (michsco) [inline Palindrome.UnitTests.cs]: This entire method is being repeated in every method. Why not make a helper method in this class:

```csharp
private void RunIsArrayPalindromeTest(string input, bool expectedOutput)
{
    Palindrome unit = new Palindrome();
    bool actualResult = unit.IsArrayPalindrome(input);
    Array.AreEqual(expectOutput, actualResult, $"{nameof(unit.IsArrayPalindrome)}(\"{input}\")");
}
```
You can then use the helper method to make your code easier to read.

File: Palindrome.UnitTests.cs [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+            char[] arrToCheck = check.PalindromeHelper("radar");
+
+            bool boolResultOfPalindromeCheck = check.IsArrayPalindrome(arrToCheck);
+
+            Assert.AreEqual(true, boolResultOfPalindromeCheck);
+        }
+        [TestMethod]
+        public void Check_Odd_Palindrome_Two()
```
Comment (michsco) [inline Palindrome.UnitTests.cs]: There's no need for two different tests since they are testing the exact same scenario (true for both even and odd).

File: Palindrome.UnitTests.cs [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+            // Arrange
+            Palindrome check = new Palindrome();
+            char[] arrToCheck = check.PalindromeHelper("");
+            // Act
+            bool boolResultOfPalindromeCheck = check.IsArrayPalindrome(arrToCheck);
+            // Assert
+            Assert.AreEqual(false, boolResultOfPalindromeCheck);
+        }
+        [TestMethod]
+        public void Check_Longest_Palindrome_In_Dictionary()
```
Comment (michsco) [inline Palindrome.UnitTests.cs]: Is there a reason (based on the code) that you need to test the longest word?

**Review by michsco — COMMENTED** [review]
I have added comments the same way I would if I was reviewing for work. Keep in mind this is an interview question, so you should timebox how much time you spend just on this.

File: Palindrome.cs (hunk 1 above)
Comment (hDoubleE) [reply Palindrome.cs]: I made everything public for consistency throughout the code, laziness really. I'll focus more on access modifiers and using them correctly. Will change.

File: Palindrome.cs (hunk 2 above)
Comment (hDoubleE) [reply Palindrome.cs]: The idea is two fold:
1: ToLower allows case insensitive, ie. Names. (Otto, Hannah)
2. Removing all spaces from sentences and replacing with "", allows input of palindromic sentences, ie.”Murder for a jar of red rum”. The idea is to smash the sentence and remove all spaces. Maybe it's no longer really a palindrome in that case. Test for this case or get rid of it?

File: Palindrome.cs (hunk 4 above)
Comment (hDoubleE) [reply Palindrome.cs]: I went back and read up on sealing classes and methods, thanks.

File: Palindrome.cs (hunk 5 above)
Comment (hDoubleE) [reply Palindrome.cs]: I'll try it.

File: Palindrome.UnitTests.cs (hunk 8 above)
Comment (hDoubleE) [reply Palindrome.UnitTests.cs]: Is this something you do in code generally? Null check everything?

File: Palindrome.UnitTests.cs (hunk 10 above)
Comment (hDoubleE) [reply Palindrome.UnitTests.cs]: True, I'll move it to test class. Sloppy design.

File: Palindrome.UnitTests.cs (hunk 11 above)
Comment (hDoubleE) [reply Palindrome.UnitTests.cs]: This is where I needed input and clarification on tests and helper methods. Will change.

File: Palindrome.UnitTests.cs (hunk 12 above)
Comment (hDoubleE) [reply Palindrome.UnitTests.cs]: Will remove redundant test cases and useless code.

File: Palindrome.UnitTests.cs (hunk 13 above)
Comment (hDoubleE) [reply Palindrome.UnitTests.cs]: Not at all, wikipedia got me reading about palindromes. Will remove.

**Comment by hDoubleE** [comment]
I will time-box this as you pointed out. The value here is really in the general tips that apply broadly. I'll implement a few changes over the weekend. Couple follow-up questions above. I'll try not to bug you and close this out. Thanks for the great feedback.

File: Palindrome.UnitTests.cs (hunk 8 above)
Comment (rileymckenna) [reply Palindrome.UnitTests.cs]: yep! check out the ?? operator and ?: operator in C# 
Imagine you request information from the server using Ajax, and you expected the value to populate your web page...but the server is slow to respond so when the page gets loaded each value is null. How do you handle those situations when you are expecting data, but an unreliable system might not have retrieved that data yet? Graceful exception? Try Catch? Null check? await/async?

File: Palindrome.UnitTests.cs (hunk 13 above)
Comment (rileymckenna) [reply Palindrome.UnitTests.cs]: dont remove it. Its another test isnt it? you did the work to make it, id keep it.

File: Palindrome.cs (hunk 2 above)
Comment (michsco) [reply Palindrome.cs]: I'd recommend handling casing and whitespace in your loop to just skip those things. You can introduce a case insensitive comparison with something like this:

```csharp
private static bool AreEqualCaseInsensitive(char left, char right)
{
    if (left == right)
    {
        return true;
    }
    else if (char.IsLetter(left) && char.IsLetter(right))
    {
        return char.ToLower(left) == char.ToLower(right);
    }

    return false;
}
```
The difference between this and the string method you have is that we aren't actually creating a new string. It is just a small math calculation to change a single character to its lower case equivalent.

> **SIDE NOTE:** Casing can bite you depending on whether the input is ASCII or Unicode. There are certain languages where there is a direct upper/lower case equivalent for certain letters. In short, casing is something to keep in mind.

File: Palindrome.UnitTests.cs (hunk 8 above)
Comment (michsco) [reply Palindrome.UnitTests.cs]: All of @rileymckenna's suggestions are always valid.

> **RULE OF THUMB:** Don't trust inputs to any public methods unless you are guaranteed about the format for some reason.

If you have a method that accepts `int`, you know it *must* be a number, but you have no guarantees about the range of the number. This is why you must ensure that you are validating parameters before making assumptions about them. Aside from the stability of your code, you also want to make sure data is validated for security reasons (e.g. SQL injection attacks).

File: Palindrome.UnitTests.cs (hunk 13 above)
Comment (michsco) [reply Palindrome.UnitTests.cs]: A few counterpoints to @rileymckenna's suggestion of keeping:

1. For a method like this, there isn't anything special about the longest valid English palindrome. You could easily generate a long palindrome from random data using letters if you wanted to check how the method behaves for large inputs.
2. You want to take equivalence classes into account when writing tests. If you have two tests that will hit the exact same code path and aren't testing a new scenario, you are just adding extra code you need to maintain going forward. This is one of the reasons that people use random input (e.g. *any* even length string with random values that are a palindrome) when running a test so that code doesn't rely on the precise values. This falls back into my "be kind to your future self" rule.

All of this said, there's definitely room for debate, especially in a scenario where you are learning to test -- you did put the work in.

File: Palindrome.cs (hunk 4 above)
Comment (michsco) [reply Palindrome.cs]: It's one of my pet peeves based on my experiences building client libraries. It's the same reason that I harp on access modifiers.

**Review by michsco — COMMENTED** [review]
Responded to comments.

**Comment by hDoubleE** [comment]
• [Turn off email notifications here...](https://help.github.com/articles/choosing-the-delivery-method-for-your-notifications/#choosing-the-delivery-method-for-notifications-about-repository-activity)tried to cut down on that with this large block.

• I did my best to incorporate all the concepts you outlined. I basically re-wrote it and probably spent too much time, but it wasn’t really about the silly palindrome program and more about the tests, helper methods, and general programming practices.

• Some of the features you laid out, ie. RunTestMethod helper, checking for case insensitivity in-place were pretty mind-blowing, so I had to dig in and ran with it.

• Based on this comment:
>This is one of the reasons that people use random input (e.g. any even length string with random values that are a palindrome)

I automated a random palindrome generator for generating large random inputs during tests.

• I also made the IsPalindrome method accept string param instead of char array. They’re essentially the same, and I could have both, but logically if I were checking if something is palindromic, I’d expect a string input. The code is basically identical, but cuts out some of the conversion logic. As mentioned:
>You don't really need your helper method.

• Thanks for everyone’s time and constructive input on this.

File: Palindrome.UnitTests.cs [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+        *  Test class contains 3 helper methods.
+        *  Passing 6/6 test cases.
+        */
+
+        /// <summary>
+        /// Generates a random string of upper and lower case letters including ' '.
+        /// Passes result string to RandomPalindromeGenerator.
+        /// </summary>
+        /// <returns>A string containing less than 1000 characters.</returns>
+        private static string RandomString()
```
Comment (michsco) [inline Palindrome.UnitTests.cs:20]: Private methods are typically placed at the end of the class.

File: Palindrome.UnitTests.cs [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+using System.Text;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Palindrome.UnitTests
+{
+    [TestClass]
+    public sealed class PalindromeUnitTests
+    {
+        /* 
+        *  Test class contains 3 helper methods.
```
Comment (michsco) [inline Palindrome.UnitTests.cs:11]: This comment is unnecessary since tests aren't always guaranteed to pass.

File: Palindrome.UnitTests.cs [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+        */
+
+        /// <summary>
+        /// Generates a random string of upper and lower case letters including ' '.
+        /// Passes result string to RandomPalindromeGenerator.
+        /// </summary>
+        /// <returns>A string containing less than 1000 characters.</returns>
+        private static string RandomString()
+        {
+            Random rand = new Random();
```
Comment (michsco) [inline Palindrome.UnitTests.cs:22]: Due to how random numbers are seeded, you should create a single instance of the `Random` class at the top of the class. I can explain more at another time.

```csharp
private readonly Random Random = new Random();
```

File: Palindrome.UnitTests.cs [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+            Palindrome unit = new Palindrome();
+            // Act
+            bool actualResult = unit.IsPalindrome(input);
+            // Asert 
+            Assert.AreEqual(expectedOutput, actualResult, $"{nameof(unit.IsPalindrome)}(\"{input}\")");
+        }
+
+        [TestMethod]
+        [ExpectedException(typeof(ArgumentNullException))]
+        public void Test_Null_Input()
```
Comment (michsco) [inline Palindrome.UnitTests.cs:95]: Why not use the helper method?

File: Palindrome.UnitTests.cs [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+            // Act
+            unit.IsPalindrome(null);
+            // Assert: Expects Exception.
+        }
+
+        [TestMethod]
+        [ExpectedException(typeof(ArgumentNullException))]
+        public void Test_Empty_Input()
+        {
+            // Arrange
```
Comment (michsco) [inline Palindrome.UnitTests.cs:108]: Why not use the helper method?

File: Palindrome.UnitTests.cs [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+        }
+
+        [TestMethod]
+        public void Test_Sentence_With_Random_Caps()
+        {
+            RunIsPalindromeTest("Murder for a JAR of reD rum", true);
+        }
+
+        [TestMethod]
+        public void Test_False_Output()
```
Comment (michsco) [inline Palindrome.UnitTests.cs:134]: This name is confusing -- what is it testing? Perhaps something like 'IsPalindrome_NonPalindrome_OddLetters`

Comment (michsco) [inline Palindrome.UnitTests.cs:134]: You probably need one for even characters as well. Perhaps using the random string.

File: Palindrome.cs [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+        /// <summary>
+        /// Method checks if input string is a palindrome.
+        /// Altered method to accept string input instead of char Array.
+        /// Functionally they are equivalent.
+        /// Modified to execute without duplicating strings using 
+        /// minimum space complexity.
+        /// </summary>
+        /// <param name="arr">A character array.</param>
+        /// <returns>Boolean result</returns>
+        public bool IsPalindrome(string strInput)
```
Comment (michsco) [inline Palindrome.cs:16]: Typically, you don't want to include the data type in the string (i.e. `str`). Input is fine.

File: Palindrome.cs [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+
+            int left = 0;
+            int right = strInput.Length - 1;
+
+            while (left < right)
+            {
+                // Skip over spaces to allow for sentences.
+                if (strInput[left] == ' ')
+                {
+                    left++;
```
Comment (michsco) [inline Palindrome.cs:32]: You should add a test case for a string "bob " (notice the space at the end). I think you'll need a "continue;" here.

**Review by michsco — COMMENTED** [review]
You have one bug related to whitespace and a few stylistic issues, but otherwise, you have answered the problem. Recommend moving on to others :-)
